src/arithmetic/rewrite_simplify.cc

```diff
@@ -626,6 +671,596 @@ Mutate_(const Mod* op, const Expr& self) {
+                       c1.Eval()->value + 1 == c2.Eval()->value);
+    TVM_TRY_REWRITE_IF(min(((x + c1) / c2) * c2, max(x, c2)), max(x, c2),
+                       c2.Eval()->value > 0 &&
+                       c1.Eval()->value + 1 == c2.Eval()->value &&
+                       CanProveGreaterEqual(x.Eval(), 0));
+
+    TVM_TRY_REWRITE_IF(min(x, ((x + c1) / c2) * c2), x,
+                       c2.Eval()->value > 0 &&
+                       c1.Eval()->value + 1 == c2.Eval()->value);
+    TVM_TRY_REWRITE_IF(min(max(x, c2), ((x + c1) / c2) * c2), max(x, c2),
+                       c2.Eval()->value > 0 &&
+                       c1.Eval()->value + 1 == c2.Eval()->value &&
+                       CanProveGreaterEqual(x.Eval(), 0));
+
+    TVM_TRY_REWRITE(min(max(x, y), min(x, y)), min(x, y));
+    TVM_TRY_REWRITE(min(max(x, y), min(y, x)), min(x, y));
+
+    TVM_TRY_REWRITE(min(max(x, y), x), x);
+    TVM_TRY_REWRITE(min(max(x, y), y), y);
+    TVM_TRY_REWRITE(min(min(x, y), x), min(x, y));
+    TVM_TRY_REWRITE(min(min(x, y), y), min(x, y));
+
+    TVM_TRY_REWRITE(min(x, max(x, y)), x);
+    TVM_TRY_REWRITE(min(y, max(x, y)), y);
+    TVM_TRY_REWRITE(min(x, min(x, y)), min(x, y));
+    TVM_TRY_REWRITE(min(y, min(x, y)), min(x, y));
+
+    TVM_TRY_REWRITE(min(min(min(x, y), z), y), min(min(x, y), z));
+    TVM_TRY_REWRITE(min(min(min(min(x, y), z), s1), y), min(min(min(x, y), z), s1));
+    TVM_TRY_REWRITE(min(min(min(min(min(x, y), z), s1), s2), y),
+                    min(min(min(min(x, y), z), s1), s2));
+
+    TVM_TRY_REWRITE(min(max(x, y), max(x, z)), max(min(y, z), x));
+    TVM_TRY_REWRITE(min(max(x, y), max(z, x)), max(min(y, z), x));
+    TVM_TRY_REWRITE(min(max(y, x), max(x, z)), max(min(y, z), x));
+    TVM_TRY_REWRITE(min(max(y, x), max(z, x)), max(min(y, z), x));
+
+    TVM_TRY_REWRITE(min(min(x, y), min(x, z)), min(min(y, z), x));
+    TVM_TRY_REWRITE(min(min(x, y), min(z, x)), min(min(y, z), x));
+    TVM_TRY_REWRITE(min(min(y, x), min(x, z)), min(min(y, z), x));
+    TVM_TRY_REWRITE(min(min(y, x), min(z, x)), min(min(y, z), x));
+
+    TVM_TRY_REWRITE(min(y + x, z + x), min(y, z) + x);
+    TVM_TRY_REWRITE(min(y + x, x + z), min(y, z) + x);
+    TVM_TRY_REWRITE(min(x + y, x + z), min(y, z) + x);
+    TVM_TRY_REWRITE(min(x + y, z + x), min(y, z) + x);
+
+    // constant folding rule.
+    TVM_TRY_REWRITE(min(min(x, c1), c2), min(x, min(c1, c2)));
+
+    // scaling rule
+    if (min(x / c1, y / c1).Match(ret)) {
+      if (c1.Eval()->value > 0) {
+        return (min(x, y) / c1).Eval();
+      } else {
+        return (max(x, y) / c1).Eval();
+      }
+    }
+    if (min(x * c1, y * c1).Match(ret)) {
+      if (c1.Eval()->value > 0) {
+        return (min(x, y) * c1).Eval();
+      } else {
+        return (max(x, y) * c1).Eval();
+      }
+    }
+    if (min(x * c1, c2).Match(ret)) {
+      int64_t c1val = c1.Eval()->value;
+      int64_t c2val = c2.Eval()->value;
+      if (c2val % c1val == 0) {
+        if (c2val / c1val > 0) {
```

Comment:
I think `>` should be `>=` here since c2val can be zero.

Reply:
updated